nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-cluster/src/main/java/org/apache/nifi/cluster/coordination/heartbeat/CuratorHeartbeatMonitor.java

```diff
@@ -0,0 +1,376 @@
+        } catch (final NoNodeException nne) {
+            logger.info("Could not find any heartbeats in ZooKeeper because the ZNode " + heartbeatPath + " does not exist");
+            return null;
+        } catch (final Exception e) {
+            logger.error("Failed to obtain heartbeats from ZooKeeper due to {}", e);
+            logger.error("", e);
+            clusterCoordinator.reportEvent(null, Severity.WARNING, "Failed to obtain heartbeats from ZooKeeper due to " + e);
+
+            return null;
+        }
+
+        logger.debug("Found {} nodes that have emitted heartbeats to ZooKeeper", nodeIds.size());
+        final Map<NodeIdentifier, NodeHeartbeat> latestHeartbeats = new HashMap<>(nodeIds.size());
+        for (final String nodeId : nodeIds) {
+            final HeartbeatMessage heartbeatMsg;
+
+            final String nodeHeartbeatPath = heartbeatPath + "/" + nodeId;
+            final Stat stat = new Stat();
+            try {
+                final byte[] serializedHeartbeatMsg = getClient().getData().storingStatIn(stat).forPath(nodeHeartbeatPath);
+                heartbeatMsg = (HeartbeatMessage) unmarshaller.unmarshal(new ByteArrayInputStream(serializedHeartbeatMsg));
+            } catch (final Exception e) {
+                logger.error("Failed to obtain heartbeat from ZooKeeper for Node with ID {} due to {}", nodeId, e);
+                logger.error("", e);
+                clusterCoordinator.reportEvent(null, Severity.WARNING, "Failed to obtain heartbeat for Node with ID " + nodeId + " from ZooKeeper due to " + e);
+
+                continue;
+            }
+
+            // update timestamp to be the timestamp that ZooKeeper reports
+            final long lastModifiedTime = stat.getMtime();
+            if (lastModifiedTime > latestHeartbeatTime) {
+                latestHeartbeatTime = lastModifiedTime;
+            }
+
+            latestHeartbeats.put(heartbeatMsg.getHeartbeat().getNodeIdentifier(), StandardNodeHeartbeat.fromHeartbeatMessage(heartbeatMsg, lastModifiedTime));
+            logger.debug("Received heartbeat from Node {}", nodeId);
+        }
+
+        logger.debug("Fetched {} heartbeats from ZooKeeper", latestHeartbeats.size());
+        return latestHeartbeats;
+    }
+
+    @Override
+    public synchronized void removeHeartbeat(final NodeIdentifier nodeId) {
+        logger.debug("Deleting heartbeat for node {}", nodeId);
+        final String nodeHeartbeatPath = heartbeatPath + "/" + nodeId.getId();
+
+        latestHeartbeatMessages.remove(nodeId);
+
+        try {
+            getClient().delete().forPath(nodeHeartbeatPath);
+            logger.info("Removed heartbeat from ZooKeeper for Node {}", nodeId);
+        } catch (final NoNodeException e) {
+            // node did not exist. Just return.
+            logger.debug("Attempted to remove heartbeat for Node with ID {} but no ZNode existed at {}", nodeId, nodeHeartbeatPath);
+            return;
```

Comment:
That is correct - we are attempting to delete a heartbeat that no longer exists... so we consider it a 'success' and are simply logging at a debug level that it was already gone.